pkg/streamingpromql/types/fpoint_ring_buffer.go

```diff
@@ -237,6 +239,49 @@ func (b *FPointRingBuffer) AnyAtOrBefore(maxT int64) bool {
 	return b.points[b.firstIndex].T <= maxT
 }
 
+// ChangesAtOrBefore returns the number of changes between subsequent points in this ring buffer from the first
+// point to the point with timestamp less than or equal to maxT.
+func (b *FPointRingBuffer) ChangesAtOrBefore(maxT int64) (float64, bool) {
```

Comment:
I think this is better fit into functions/range_vectors in terms of consistency. Can we not create a test in the functions package and pass it a constructed `types.RangeVectorStepData`?

Reply:
> I think this is better fit into functions/range_vectors in terms of consistency

Actually some functions are implemented in `fpoint_ring_buffer.go` too like `CountAtOrBefore`, `LastAtOrBefore` and `AnyAtOrBefore`.

> Can we not create a test in the functions package and pass it a constructed types.RangeVectorStepData?
It is not possible to create the buffer to simulate wrap around because `points` and other fields are private. I prefer not to make them public just for testing.

Other than that, if we put the logic in `range_vectors.go` we are exposing the internal representation of the ring buffer in `range_vector.go`. Like on how iterating head and tail which I believe should be internal `fpoint_ring_buffer.go`. 

At the end I don't have any strong opinion and I am fine if there is a better argument.

Reply:
> Can we not create a test in the functions package and pass it a constructed types.RangeVectorStepData?
It is not possible to create the buffer to simulate wrap around because the fields are private.

Reply:
I would favour this not being on the ring buffer type - `ChangesAtOrBefore` can efficiently be implemented by using `UnsafePoints`, whereas `CountAtOrBefore`, `LastAtOrBefore` and `AnyAtOrBefore` don't need the slices produced by `UnsafePoints` and so benefit from being implemented directly on the ring buffer type.

Reply:
I will moved back to `range_vectors.go` then. I will figure out how the best to test that.